### DISCUSSION TITLE
Access to JTable validation check() function

### DISCUSSION THREAD

**Discussion by kevincam3** [joomengine/Joomla-Component-Builder#740] [Ideas]
One thing I can't seem to time in the Wiki/docs or the videos is how to add my own code in the JTables Check() function since I'm doing an API and need to do the validation check at the database/table level. Is that possible with JCB?

**Comment by kevincam3** [comment]
Thanks for then info about how this can be done.

I'm quoting you below: 

"Happy to hear you got that issue sorted... please do not cross issue boundaries... since other looking for answers will not find it. Next time please post a question in the https://github.com/vdm-io/Joomla-Component-Builder/discussions area as it is the new JCB forum...

Okay so the JTable is a class that we extend for the edit views which has a check method which we do not target from the JCB GUI at this time. So you will need to use some custom code..."

**Comment by kevincam3** [comment]
I don't do it a lot but if it was easier to do I might do that check more often in JTable since I think that's where that validation should live most of the time.

**Comment by kevincam3** [comment]
Sorry,

Just seeing the discussion link.

**Comment by Llewellynvdm** [comment]
Validation is done [in the controller](https://github.com/joomla/joomla-cms/blob/3.9.26/libraries/src/MVC/Controller/FormController.php#L714) via the [model](https://github.com/joomla/joomla-cms/blob/3.9.26/libraries/src/MVC/Model/FormModel.php#L357), but with the [JForm class](https://github.com/joomla/joomla-cms/blob/3.9.26/libraries/src/MVC/Model/FormModel.php#L228) returned by the [getForm Method](https://github.com/vdm-io/Joomla-Component-Builder/blob/staging/admin/models/field.php#L309) so that is why we have access to this area via JCB GUI.

![image](https://user-images.githubusercontent.com/5607939/116948525-dc31b000-ac7f-11eb-8153-f6c63298fce7.png)

So I don't see why I would want to move my validation to the table, as this is after the fact... where the from validation is triggered before, the table. The table is closer to the database... so the way it validates the data now in the MVC we rely on the field validation. So when you setup your fields you can add custom validation to any field.

Then with the permissions JCB has per field, if you say a user group does not have access to submit a field... then this is validated in the getForm Method automatically by JCB. Should you have a pro membership plugin, this validation goes even down into the Model with export and import of data. So with the combination of field validation and user access control I would say we have a very secure system without the need to touch the table.

Let me know if you have any questions, or if I need to do some recording on this explanation.

**Reply by kevincam3** [reply]
Thanks for that information. The reason that doesn't work in my situation is because it's an API interface for Joomla, it doesn't have any models. The flow is Contoller to Plugin and in the plugin I'm using JTable because all I'm doing is saving and/or updating rows in just one table. So using JTable is nice because I don't have to write any queries at all. 

But that's why I'm using the check() function in JTable to validate the data since there's already that check happening in JTable for other things. So being able to check there and just return false let's me do this.

		// save the data
		if (!$table->save($data))
		{

			ApiError::raiseError(400, "Data Didn't Save!", 'APIException');
		} else {
		$this->plugin->setResponse('success');
		}
		
The API Component I'm using is this one. https://github.com/techjoomla/com_api for reference.

**Reply by Llewellynvdm** [reply]
But none of your field validation is there... remember that. So I would say the API should work via the model, and no directly with the table class. But that is me being lazy.. I like to keep concern centralized. Meaning instead of maintaining two models... moving modeling into the Table would kind of be that kind of effect... I would instead get the model and save items and get items that way... that is the current path anyway.

This allows me to forget about the validation... since it is always in the works. I have done this so many times I actually have a planned multi form substitution script. This allows me to bring multiple forms into one form... and then post back to multiple tables.

Little bit of tweaking to the controller and a little helper method and we're done. API is not that hard to setup with the current flow... there has been talk amongst the JCB developers to integrate https://www.selfget.com/products/jbackend.html

**Comment by kevincam3** [comment]
Yeah, I hear what you're saying. I've heard about it jbackend but never used it. Since J4 will have a built in API I'm looking forward to JCB working with it and will probably use it in the long run.

I ended up using the Custom Code to edit the JTable Check() function and it worked out fine thanks for reminding me about that option.
